Added Authenticated download-and-unzip installer

pjdarton opened the pull request:
Much as the core download-and-unpack method, but this supports basic authentication on the download.
With increasing focus on security, HTTP is giving way to HTTPS and anonymous is giving way to authenticated access.  This allows custom tools etc to be downloaded from servers that do not permit anonymous downloads.

oleg-nenashev commented on src/main/java/com/synopsys/arc/jenkinsci/plugins/extratoolinstallers/installers/AuthenticatedDownloadCallable.java:
```diff
@@ -0,0 +1,283 @@
+class AuthenticatedDownloadCallable extends MasterToSlaveFileCallable<Date> {
+    private static final long serialVersionUID = 1L;
+    private final URI uri;
+    private final String usernameOrNull;
```
`@CheckForNull` would help here and below

oleg-nenashev commented on src/main/java/com/synopsys/arc/jenkinsci/plugins/extratoolinstallers/installers/AuthenticatedZipExtractionInstaller.java:
```diff
@@ -0,0 +1,560 @@
+            if (item == null) {
+                return Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER);
+            }
+            return item.hasPermission(Item.EXTENDED_READ) || item.hasPermission(CredentialsProvider.USE_ITEM);
```
I doubt it ever possible for Tool Installers. They should be configurable only by admins, and it may be better to hardcode it

pjdarton replied:
OK, I've removed `item` and the associated complexity.

oleg-nenashev commented on src/main/java/com/synopsys/arc/jenkinsci/plugins/extratoolinstallers/installers/AuthenticatedZipExtractionInstaller.java:
```diff
@@ -0,0 +1,560 @@
+    }
+
+    @Extension
+    public static class DescriptorImpl extends ToolInstallerDescriptor<AuthenticatedZipExtractionInstaller> {
```
`@Symbol(`something`)`

pjdarton commented:
I've...
 - addressed the review comments.
 - acted on [your suggestion on my other PR](https://github.com/jenkinsci/extra-tool-installers-plugin/pull/8#discussion_r249211899) to use a different package name.
 - re-based this PR on the current master
 - removed some debug logging I'd accidentally left in (it logged the request headers and response headers)
 - changed the code to use a HTTP HEAD request when validating the form configuration as, even if ignoring the response contents, HTTP GET took ages to validate for large files over slow links.

pjdarton commented:
I think that the core code should support authenticated downloads - it seems quite unreasonable to me that 99% of all of Jenkins plugin functionality supports providing credentials to external servers, but when it comes to downloading tools from a zip/tar.gz it's anonymous-only because the core FilePath code doesn't support authenticated downloads.
It would be nice if this kind of functionality made its way into the core code ... but until then, it can live here in this plugin :grin:

oleg-nenashev commented:
@pjdarton there are some trivial merge conflicts after the #8 merge, can be fixed easily

pjdarton commented:
Sure thing; I expected as much.  Messages.properties I'd guess.  I'll sort it out tomorrow.

I hope to be in a position to deploy and use this code "for real" to workaround issues at work "soon" (aka "not soon enough" as far as my developers are concerned) so I think that'll give the plugin some "real world testing" and, assuming no further changes are required, that would be a good point to do a release. WDYT?
